### PR TITLE
ENH: add ignore index to DataFrame / Series.sample

### LINKED ISSUE #38581: ENH: add ignore_index to DataFrame / Series.sample
#### Is your feature request related to a problem?

When sampling a dataframe / series we get the original index back by default, which is good, but we should add an option to ignore the index to keep it consistent with other methods like `drop_duplicates` and `explode` for example.
#### Describe the solution you'd like

Add `ignore_index` argument to Series/DataFrame.sample.
#### API breaking implications

None as far as I can tell.

```python
df = pd.DataFrame(
    {"col1": range(10, 20), "col2": range(20, 30), "colString": ["a"] * 10}
)
df.sample(n=3, ignore_index=True)

   col1  col2 colString
0    14    24         a
1    17    27         a
2    11    21         a
```

### PR DESCRIPTION
**Pull request by erfannariman**
- [x] closes #38581
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -309,6 +309,7 @@ Other enhancements
 - Implement method ``cross`` for :meth:`DataFrame.merge` and :meth:`DataFrame.join` (:issue:`5401`)
 - When :func:`read_csv/sas/json` are called with ``chuncksize``/``iterator`` they can be used in a ``with`` statement as they return context-managers (:issue:`38225`)
 - Augmented the list of named colors available for styling Excel exports, enabling all of CSS4 colors (:issue:`38247`)
+- :meth:`DataFrame.sample` now accepts the ``ignore_index`` argument to reset the index after sampling, similar to :meth:`DataFrame.drop_duplicates` and `DataFrame.sort_values` (:issue:`38581`)
```
Comment (arw2019) [inline doc/source/whatsnew/v1.2.0.rst:312]: I would expect this to go into 1.3

Comment (erfannariman) [reply doc/source/whatsnew/v1.2.0.rst]: Sure, moved it.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -5134,12 +5134,15 @@ def tail(self: FrameOrSeries, n: int = 5) -> FrameOrSeries:
...
-        random_state=None,
-        axis=None,
+        n: Optional[int] = None,
+        frac: Optional[float] = None,
+        replace: bool_t = False,
+        weights: Union[str, np.ndarray] = None,
+        random_state: Optional[
+            Union[int, arraylike, np.random.BitGenerator, np.random.RandomState]
+        ] = None,
+        axis: Union[int, str] = None,
```
Comment (MarcoGorelli) [inline pandas/core/generic.py:5144]: There is already
```python
Axis = Union[str, int]
```
in `_typing.py`

Also, I guess `arraylike` was meant to be `ArrayLike`?

Comment (arw2019) [reply pandas/core/generic.py:5144]: Also this would need to be `Optional[Axis]` since can be `None`

File: pandas/core/generic.py [hunk 3]
```diff
@@ -5134,12 +5134,15 @@ def tail(self: FrameOrSeries, n: int = 5) -> FrameOrSeries:
...
-        n=None,
-        frac=None,
-        replace=False,
-        weights=None,
-        random_state=None,
-        axis=None,
+        n: Optional[int] = None,
+        frac: Optional[float] = None,
+        replace: bool_t = False,
+        weights: Union[str, np.ndarray] = None,
```
Comment (arw2019) [inline pandas/core/generic.py:5140]: needs to be `'Optional`

Also could this be an ExtensionArray (in which case use `ArrayLike`)

File: pandas/core/generic.py (hunk 2 above)
Comment (erfannariman) [reply pandas/core/generic.py]: Thanks for the feedback, made the adjustments.

File: pandas/core/generic.py (hunk 3 above)
Comment (erfannariman) [reply pandas/core/generic.py]: Thanks, fixed it in last commit.

File: pandas/core/generic.py [hunk 4]
```diff
@@ -5134,12 +5135,15 @@ def tail(self: FrameOrSeries, n: int = 5) -> FrameOrSeries:
...
-        replace=False,
-        weights=None,
-        random_state=None,
-        axis=None,
+        n: Optional[int] = None,
+        frac: Optional[float] = None,
+        replace: bool_t = False,
+        weights: Optional[Union[str, ArrayLike]] = None,
+        random_state: Optional[
+            Union[int, ArrayLike, np.random.BitGenerator, np.random.RandomState]
```
Comment (jreback) [inline pandas/core/generic.py:5143]: move this definition to _typing (call it RandomState i think is ok) or RandomStateType if it conflicts)

**Review by jreback — CHANGES_REQUESTED** [review]
looks good small comment. pls merge master (and some checks are failing)

File: pandas/_typing.py [hunk 5]
```diff
@@ -158,3 +158,6 @@
 ColspaceArgType = Union[
     str, int, Sequence[Union[str, int]], Mapping[Label, Union[str, int]]
 ]
+
+# random generator
+RandomState = Optional[Union[int, ArrayLike, np.random.BitGenerator, np.random.RandomState]]
```
Comment (MarcoGorelli) [inline pandas/_typing.py:163]: perhaps make `RandomState = Union[...` and then use `random_state: Optional[RandomState] = None,` in generic.py

Comment (jreback) [reply pandas/_typing.py:163]: agreed, this shouldn't be Optional (that can be in the signature itself)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by erfannariman** [comment]
Not sure why this fails in CI:

```
E   AttributeError: module 'numpy.random' has no attribute 'Generator'
```

Locally this seems to work:

```python
In [1]: import numpy as np

In [2]: from typing import Union

In [3]: np.random.Generator
Out[3]: numpy.random._generator.Generator

In [4]: RandomState = Union[np.random.Generator, np.random.RandomState]
```

**Comment by arw2019** [comment]
> Not sure why this fails in CI:
> 
> ```
> E   AttributeError: module 'numpy.random' has no attribute 'Generator'
> ```
> 
> Locally this seems to work:
> 
> ```python
> In [1]: import numpy as np
> 
> In [2]: from typing import Union
> 
> In [3]: np.random.Generator
> Out[3]: numpy.random._generator.Generator
> 
> In [4]: RandomState = Union[np.random.Generator, np.random.RandomState]
> ```

This looks relevant: https://github.com/aleju/imgaug/issues/537

**Comment by erfannariman** [comment]
> This looks relevant: [aleju/imgaug#537](https://github.com/aleju/imgaug/issues/537)

Might be, but then again that topic looks quite old, and I am using the latest numpy version (1.19.4)

**Comment by arw2019** [comment]
> 
> 
> > This looks relevant: [aleju/imgaug#537](https://github.com/aleju/imgaug/issues/537)
> 
> 
> 
> Might be, but then again that topic looks quite old, and I am using the latest numpy version (1.19.4)

Right and it works for you but what's the numpy version on the failing CI task

**Comment by erfannariman** [comment]
> Right and it works for you but what's the numpy version on the failing CI task

Yup, you are probably right, we are using a version 1.16.6, didnt expect that to be honest.

https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=50646&view=logs&j=4b4d380c-92b5-58cf-5b80-2c6c9b1d5805&t=875b853d-e31b-5e7e-0c21-cef3db7ce4e5&l=123

**Comment by arw2019** [comment]
> > Right and it works for you but what's the numpy version on the failing CI task
> 
> 
> 
> Yup, you are probably right, we are using version 1.16.6, didnt expect that to be honest.
> 
> 
> 
> https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=50646&view=logs&j=4b4d380c-92b5-58cf-5b80-2c6c9b1d5805&t=875b853d-e31b-5e7e-0c21-cef3db7ce4e5&l=123

Ok! You can use np_version_under_1p17 from pandas.compat.numpy to check for that and set the alias to be Any (or stricter if possible) in that case

**Comment by erfannariman** [comment]
> Ok! You can use np_version_under_1p17 from pandas.compat.numpy to check for that and set the alias to be Any (or stricter if possible) in that case

Thanks, didn't know about `np._version_under_1p17`. To be strict we can do:

```python
if np_version_under1p17
    RandomState = Union[int, ArrayLike,  np.random.RandomState]
else:
    RandomState = Union[int, ArrayLike, np.random.Generator, np.random.RandomState]
```
But not sure if this is desirable design wise. Else  I think the only option left is to use `typing.Any`

**Comment by arw2019** [comment]
> > Ok! You can use np_version_under_1p17 from pandas.compat.numpy to check for that and set the alias to be Any (or stricter if possible) in that case
> 
> Thanks, didn't know about `np._version_under_1p17`. To be strict we can do:
> 
> ```python
> if np_version_under1p17:
>     RandomState = Union[int, ArrayLike, np.random.RandomState]
> else:
>     RandomState = Union[int, ArrayLike, np.random.Generator, np.random.RandomState]
> ```
> 
> But not sure if this is desirable design wise. Else I think the only option left is to use `typing.Any`

I don't have an opinion here - I'm not sure how much we care about making type checking backward compatible with old versions of numpy. I'd say for now push either fix to get to green and @jreback can advise since o/w this PR is good to go

**Comment by erfannariman** [comment]
@jreback @arw2019 the failing CI seems unrelated, let me know if anything else has to be done for this PR.

**Comment by MarcoGorelli** [comment]
Are you sure?
```
pandas/_typing.py:169: error: Cannot assign multiple types to name "RandomState" without an explicit "Type[...]" annotation [misc]
```
doesn't seem unrelated

**Comment by erfannariman** [comment]
> Are you sure?
> 
> ```
> pandas/_typing.py:169: error: Cannot assign multiple types to name "RandomState" without an explicit "Type[...]" annotation [misc]
> ```
> 
> doesn't seem unrelated. See [here](https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases) for some help on that

Apparently this error is raised because I define `RandomState` twice in an if else. When I remove the if else, there is no such exception.

**Comment by erfannariman** [comment]
@MarcoGorelli  @arw2019 @jreback for now I removed defining the `RandomState` type and set it to `Any`. After moving all numpy versions in the tests >= 1.17 we can add:

```python
RandomState = Union[int, ArrayLike, np.random.Generator, np.random.RandomState]
```

**Comment by jreback** [comment]
you can just define the type conditionally based on numpy version 

do not use Any (u can also remove the Any and open an issue)

**Comment by erfannariman** [comment]
> you can just define the type conditionally based on numpy version

Can you check my comment above about: `if np_version_under1p17:`, that does not seem to work. 

> do not use Any (u can also remove the Any and open an issue)

You mean an issue that we still have to set typehint for `RandomState` I assume?

**Comment by jreback** [comment]
a number of typing error: https://github.com/pandas-dev/pandas/pull/38594/checks?check_run_id=1631115163

**Comment by MarcoGorelli** [comment]
Fixing the typing here isn't so easy - there's

https://github.com/pandas-dev/pandas/blob/a97d90ab8b9497a86491167e031af00567c1e42c/pandas/core/generic.py#L5307

which probably wants renaming, because on the right you have `Union[str, ArrayLike]` and you're trying to cast it to `Series`.

Then there's

https://github.com/pandas-dev/pandas/blob/a97d90ab8b9497a86491167e031af00567c1e42c/pandas/core/generic.py#L5321

, where the return type of `Series.fillna` can be `Series` or `None` (depending on the value of `inplace`, but `mypy` doesn't know that). So I think either `fillna` needs an overload so that when `inplace` is `Literal[False]` the return type is `Series`, or we can just use `cast` (cc @simonjayhawkins )

**Comment by MarcoGorelli** [comment]
Looks like the overload approach is taken in `reset_index`

https://github.com/pandas-dev/pandas/blob/91595138a33b4763942f0612e6c2bec163e15d51/pandas/core/frame.py#L4841-L4872

**Comment by erfannariman** [comment]
> Fixing the typing here isn't so easy - there's
> 
> https://github.com/pandas-dev/pandas/blob/a97d90ab8b9497a86491167e031af00567c1e42c/pandas/core/generic.py#L5307
> 
> which probably wants renaming, because on the right you have `Union[str, ArrayLike]` and you're trying to cast it to `Series`.
> 
> Then there's
> 
> https://github.com/pandas-dev/pandas/blob/a97d90ab8b9497a86491167e031af00567c1e42c/pandas/core/generic.py#L5321
> 
> , where the return type of `Series.fillna` can be `Series` or `None` (depending on the value of `inplace`, but `mypy` doesn't know that). So I think either `fillna` needs an overload so that when `inplace` is `Literal[False]` the return type is `Series`, or we can just use `cast` (cc @simonjayhawkins )

Thanks for having a in-depth look at this. This turned out to be way more complex and time consuming than I thought. Would it be an idea to open a new ticket for overload on `fillna`?

**Comment by MarcoGorelli** [comment]
Sounds reasonable to me. If this is ready apart from that, maybe it's OK to leave `weights` untyped here for this PR, and take care of it elsewhere?

**Comment by jreback** [comment]
@erfannariman if you can merge master and see if can get to green (we can defer the typing on the random if that's causing an issue)

**Comment by erfannariman** [comment]
> @erfannariman if you can merge master and see if can get to green (we can defer the typing on the random if that's causing an issue)

Yes sure, will do this weekend, this one is basically ready to go except the typing complexity.

**Comment by MarcoGorelli** [comment]
> Thanks for having a in-depth look at this. This turned out to be way more complex and time consuming than I thought. Would it be an idea to open a new ticket for overload on `fillna`?

This is done now, if you merge upstream/master then the typing should be a lot simpler here

**Comment by simonjayhawkins** [comment]
@erfannariman can you merge master https://github.com/pandas-dev/pandas/pull/38594#issuecomment-822760699

**Comment by simonjayhawkins** [comment]
@erfannariman Thanks for the PR. closing as stale. ping if you want to continue.

**Pull request closed without merging**